src/wayland/shortcuts_inhibit/inhibitor.cpp

```diff
@@ -0,0 +1,180 @@
+#include "inhibitor.hpp"
+
+#include <private/qwaylandwindow_p.h>
+#include <qlogging.h>
+#include <qobject.h>
+#include <qtmetamacros.h>
+
+#include "../../window/proxywindow.hpp"
+#include "../../window/windowinterface.hpp"
+#include "proto.hpp"
+
+namespace qs::wayland::shortcuts_inhibit {
+using QtWaylandClient::QWaylandWindow;
+
+ShortcutsInhibitor::ShortcutsInhibitor() {
+	this->bBoundWindow.setBinding([this] {
+		return this->bEnabled ? this->bWindowObject.value() : nullptr;
+	});
+}
+
+ShortcutsInhibitor::~ShortcutsInhibitor() {
+	if (!this->inhibitor) return;
+
+	auto* manager = impl::ShortcutsInhibitManager::instance();
+	if (!manager) return;
+
+	QObject::disconnect(this->inhibitor, nullptr, this, nullptr);
+	manager->unrefShortcutsInhibitor(this->inhibitor);
+}
+
+QObject* ShortcutsInhibitor::window() const { return this->bWindowObject; }
+
+void ShortcutsInhibitor::setWindow(QObject* window) {
+	if (window == this->bWindowObject) return;
+
+	auto* proxyWindow = qobject_cast<ProxyWindowBase*>(window);
+
+	if (proxyWindow == nullptr) {
+		if (auto* iface = qobject_cast<WindowInterface*>(window)) {
+			proxyWindow = iface->proxyWindow();
+		}
+	}
+
+	this->bWindowObject = proxyWindow ? window : nullptr;
+}
+
+bool ShortcutsInhibitor::isActive() const {
+	return this->inhibitor ? this->inhibitor->isActive() : false;
+}
+
+void ShortcutsInhibitor::boundWindowChanged() {
+	auto* window = this->bBoundWindow.value();
+	auto* proxyWindow = qobject_cast<ProxyWindowBase*>(window);
+
+	if (proxyWindow == nullptr) {
+		if (auto* iface = qobject_cast<WindowInterface*>(window)) {
+			proxyWindow = iface->proxyWindow();
+		}
+	}
+
+	if (proxyWindow == this->proxyWindow) return;
+
+	if (this->mWaylandWindow) {
+		QObject::disconnect(this->mWaylandWindow, nullptr, this, nullptr);
+		this->mWaylandWindow = nullptr;
+		this->onWaylandSurfaceDestroyed();
+	}
+
+	if (this->proxyWindow) {
+		QObject::disconnect(this->proxyWindow, nullptr, this, nullptr);
+		this->proxyWindow = nullptr;
+	}
+
+	if (proxyWindow) {
+		this->proxyWindow = proxyWindow;
+
+		QObject::connect(
+		    proxyWindow,
+		    &QObject::destroyed,
+		    this,
+		    &ShortcutsInhibitor::onWindowDestroyed
+		);
+
+		QObject::connect(
+		    proxyWindow,
+		    &ProxyWindowBase::backerVisibilityChanged,
+		    this,
+		    &ShortcutsInhibitor::onWindowVisibilityChanged
+		);
+
+		this->onWindowVisibilityChanged();
+	}
+
+	emit this->windowChanged();
```

Comment:
It looks like `window` and `bBoundWindow` should be pretty much unrelated. As window hasn't changed here, it shouldn't update. If the window is actually connected should be transparent to the `window` property, and instead exposed through a readonly `active` property (which could be as simple as making inhibitor a bindable and defining active as inhibitor != null).